spec/index.md

```diff
@@ -670,37 +680,107 @@ See [The system state tree](#state-tree) for details on the state tree.
+
+-   (optional) `requested_signatures` (`text`): Determines whether and how many signatures are returned in the response; possible values are `"none"`, `"one"`.
+
+-   (optional) `requested_certificate` (`text`): Determines whether a certificate (as specified in [Certification](#certification)) is returned in the response; possible values are `"no"`, `"yes"`.
+
+The response is a CBOR (see [CBOR](#cbor)) with the same fields as for [Request: Query call](#http-query) response and the following **additional** fields:
+
+-   (optional) `signatures` (`[+ node-signature]`): a non-empty list of node signatures for the returned query response.
+
+-   (optional) `certificate` (`blob`): a certificate, as specified in [Certification](#certification).
+
+Unless `requested_signatures` is null or set to `"none"` in the request, the response to a signed query call contains a non-empty list of signatures for the returned response produced by the individual IC nodes that computed the same returned response. Every such signature (whose type is denoted as `node-signature`) is a CBOR (see [CBOR](#cbor)) map with the following fields:
 
-Canister methods that do not change the canister state can be executed more efficiently. This method provides that ability, and returns the canister's response directly within the HTTP response.
+-   `timestamp` (`nat`): the timestamp of the signature.
+
+-   `signature` (`blob`): the actual signature.
+
+-   `identity` (`principal`): the principal of the node producing the signature.
+
+Unless `requested_certificate` is null or set to `"no"` in the request, the response to a signed query call contains a certificate containing all paths in the state tree with prefixes
+
+-   `/time`. Always contained in the certificate.
+
+-   `/subnet/<subnet_id>/canister_ranges`. Contained in the certificate if the certificate does not have a subnet delegation, as specified in [Delegation](#certification-delegation).
+
+-   `/subnet/<subnet_id>/nodes/<node_id>/public_key`. Contained in the certificate if `<node_id>` is the principal of a node in `signatures`.
+
+where `<subnet_id>` characterizes the subnet hosting the requested canister.
+
+Given a response `R` and a certificate `Cert` that is
+
+-   contained in the response, i.e., `Cert = R.certificate`, or
+
+-   obtained by requesting the path `/subnet` in a **separate** read state request to `/api/v2/canister/<ECID>/read_state`,
+
+the following predicate describes when the returned response `R` is correctly signed by the nodes:
+
+    verify_response(R, Cert)
+      = verify_cert(Cert) ∧
+        ((Cert.delegation = NoDelegation ∧ SubnetId = RootSubnetId ∧ lookup(["subnet",SubnetId,"canister_ranges"], Cert) = Found Ranges) ∨
+         (SubnetId = Cert.delegation.subnet_id ∧ lookup(["subnet",SubnetId,"canister_ranges"], Cert.delegation.certificate) = Found Ranges)) ∧
```

Comment:
In the theory of the spec, we support nested delegations (and our agents actually go through the trouble of supporting them), while here we seem to assume that there's only a single delegation. This could make the check insecure if we ever switched to nested delegations. 

I'd be happy to kill nested delegations though unless someone has a hot use case for them.

Reply:
Even with nested delegations, the actual subnet ID will be the subnet ID of the last delegation and the canister ranges of every delegation should be included in the canister ranges of its parent delegation (this condition is currently not in the spec, but how about it we just add this condition).

Reply:
I don't think my comment about this becoming insecure made sense - sorry. But this still seems off if we have nested delegations. If the subnet changes along the chain of delegations, which subnet should we pick? This seems to pick the first one in the chain. Is that the right choice? Should it be the last one in the chain instead? Should it be any subnet in the chain? But then we probably don't want to allow mixing and matching different subnets, otherwise a selection of malicious nodes from each subnet could reach a quorum?

So I'd still propose to just kill nested delegations if we don't have a use for them, and if noone remembers what the envisioned future was where they would make sense. Right now they just seem to be making our lives more complicated.

Reply:
> If the subnet changes along the chain of delegations, which subnet should we pick?

We should pick the outermost one (as we're doing it now) because the delegation starts with NNS being the innermost.

Reply:
Regarding why we even have nested delegations: wouldn't it be more complicated to enforce we have only have a single delegation? In the verification algorithm running on a certificate, you'd need a Boolean value expressing whether it's the outer or inner certificate; now you can just define the verification recursively without focusing on the recursion depth.

Reply:
In the Interface Specification meeting, we decided to follow up on removing nested delegations separately. Hence, I'm closing this discussion.